File: 2017-12/2017-12-25/leetcode120.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        if(triangle.empty()) return 0;
        int height = triangle.size();
        int len;
        for(int i = 1; i < height; i++)
        {
            len = triangle[i].size();
            for(int j = 0; j < len; j++)
            {
                if(!j)
                {
                    triangle[i][j] += triangle[i - 1][0];
                }
                else if(j == len - 1)
                {
                    triangle[i][j] += triangle[i - 1][j - 1];
                }
                else
                {
                    triangle[i][j] += min(triangle[i - 1][j - 1],triangle[i - 1][j]);
                }
            }
        }
        int res = triangle[height - 1][0];
        for(int i = 0; i < triangle[height - 1].size(); i++)
        {
            res = min(res, triangle[height - 1][i]);
        }
        return res;
    }
};

File: 2018-02/2018-02-22/leetcode120.java
/**
 * date: 2018-02-22
 * author: pwxcoo
 * complexity: T = O(n), S = O(n)
 * describe:    dp
 */

class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        for(int i = 1; i <= triangle.size(); i++)
        {
            List<Integer> temp = triangle.get(i - 1);
            if(i == 1) continue;
            List<Integer> last = triangle.get(i - 2);
            for(int j = 0; j < temp.size(); j++)
            {
                if(j == 0)
                    temp.set(0, temp.get(0) + last.get(0));
                else if(j == temp.size() - 1)
                    temp.set(j, temp.get(j) + last.get(j - 1));
                else 
                    temp.set(j, temp.get(j) + Math.min(last.get(j - 1), last.get(j)));
            }
        }
        
        int len = triangle.size();
        int res = triangle.get(len - 1).get(0);
        for(int i = 1; i < len; i++)
        {
            res = Math.min(triangle.get(len - 1).get(i), res);
        }
        return res;
    }
}